src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Utilities/TypeStyle/CSharpUseImplicitTypeHelper.cs

```diff
@@ -381,10 +384,43 @@ private static bool IsSwitchExpressionAndCannotUseVar(TypeSyntax typeName, Expre
                         {
                             noValidTypeExpressions = false;
                             var expressionType = semanticModel.GetTypeInfo(expression).Type;
-                            if (expressionType != null && !expressionType.InheritsFromOrEquals(declarationType))
+                            if (expressionType == null)
+                            {
+                                continue;
+                            }
+
+                            if (!expressionType.InheritsFromOrEquals(declarationType))
                             {
                                 return true;
                             }
+
+                            // All arms must be in the same direct line of inheritance.
+                            // e.g. Given the tree:
+                            //   C
+                            //  / \
+                            // A   B
+                            //
+                            // Substituting 'var' for 'x' in the following switch expression will introduce a compiler error
+                            // due to the potential for ambiguity (e.g. if C inherited from another class D).
+                            // C x = i switch
+                            // {
+                            //     0 => new A(),
+                            //     1 => new B(),
+                            //     _ => throw new ArgumentException(),
+                            // };
+                            if (expressionType.Equals(declarationType) || seenTypes.Contains(expressionType))
+                            {
+                                continue;
+                            }
+
+                            var invalidType = seenTypes.Any(
+                                t => !t.InheritsFromOrEquals(expressionType) && !expressionType.InheritsFromOrEquals(t));
```

Comment:
so this fixes the bug, but i don't think it's accurate in terms of how the language sees things.  for example, if you have:

```c#
interface I { }
class C : I { }

i switch {
  0 => (C)...,
  1 => (I)...,
}
```

THen that would ok.  I'm not sure the exact rules (since i can't find the spec on it), but it's likely going to be more of 'is there a common type between these?'.

@jcouv @gafter can you point us at the spec for switch-expressions so we can implement the right check as per the language?

Reply:
Here's the spec for switch-expressions: https://github.com/dotnet/csharplang/blob/master/proposals/csharp-8.0/patterns.md#switch-expression